### PR TITLE
[Coverage][MCDC] Group mcdc tests and fix panic when generating mcdc code for inlined expressions.

### PR DESCRIPTION
**Pull request by ZhuUx**
### Changes

1. Group all mcdc tests to one directory.
2. Since mcdc instruments different mappings for boolean expressions with normal branch coverage as #125766 introduces, it would be better also trace branch coverage results in mcdc tests.
3. So far rustc does not call `CoverageInfoBuilderMethods::init_coverage` for inlined functions. As a result, it could panic if it tries to instrument mcdc statements for inlined functions due to uninitialized cond bitmaps. We can reproduce this issue by current nightly rustc and [the test](https://github.com/rust-lang/rust/pull/127234/files#diff-c81af6bf4869aa42f5c7334e3e86344475de362f673f54ce439ec75fcb5ac3e5) with flag `--release`. This patch fixes it.

### CODE REVIEW

**Comment by ZhuUx** [comment]
@rustbot label +A-code-coverage
cc @RenjiSann

File: compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs [hunk 1]
```diff
@@ -93,6 +97,9 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
 
 impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
     fn init_coverage(&mut self, instance: Instance<'tcx>) {
+        if !self.coverage_context().is_some_and(|ctx| !ctx.is_initialized_for(&instance)) {
```
Comment (ZhuUx) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:100]: This should be equivalent to `is_none_or(|ctx| ctx.is_initialized_for(&instance))`. However `is_none_or` has not been stabilized yet.

**Comment by Zalathar** [comment]
Unfortunately, I'm pretty sure this fix is wrong.

We can't just call `init_coverage` in the middle of visiting an inlined coverage statement, because (even with these changes) that method is written around the assumption that it will be called exactly once, at the *start* of the relevant function.

For example, what happens if a function is inlined in multiple places, or is both inlined and also codegenned as a non-inlined function? It looks like only some of those places will end up with the local variables that they need.

---

For now, I think the workaround should just be to completely disable MIR inlining when MC/DC is enabled.

**Comment by Zalathar** [comment]
By the way, if you want to extract the first two patches (adjusting tests) to their own PR, they should be fine.

(I can't directly approve them myself, but I can at least tell the assigned reviewer that they're OK.)

**Comment by Zalathar** [comment]
Something like this in `compiler/rustc_mir_transform/src/inline.rs` should work around the problem:

```rust
impl<'tcx> MirPass<'tcx> for Inline {
    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
        // FIXME(#127234): Coverage instrumentation currently doesn't handle
        // inlined MIR correctly when MC/DC is enabled.
        if sess.instrument_coverage_mcdc() {
            return false;
        }
```

**Comment by ZhuUx** [comment]
> We can't just call `init_coverage` in the middle of visiting an inlined coverage statement, because (even with these changes) that method is written around the assumption that it will be called exactly once, at the _start_ of the relevant function.

How about use `(RootInstance, InlinedInstance)` as the key to reference the local mcdc parameters?

**Comment by ZhuUx** [comment]
Considering there might be much overhead if we created mcdc parameters for each inlined functions separately. Disable inlining when mcdc is enabled seems better for now.

File: tests/coverage/mcdc/if.cov-map [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Function name: mcdc_if::mcdc_check_a
```
Comment (Zalathar) [inline tests/coverage/mcdc/if.cov-map:1]: Renaming the test files is fine, but I don't think we should rename the functions in the files.

It makes the history harder to follow, and there isn't any corresponding benefit.

File: compiler/rustc_mir_transform/src/inline.rs [hunk 3]
```diff
@@ -41,6 +41,12 @@ struct CallSite<'tcx> {
 
 impl<'tcx> MirPass<'tcx> for Inline {
     fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        // FIXME(#127234): Coverage instrumentation currently doesn't handle
+        // inlined MIR correctly when MC/DC is enabled.
```
Comment (Zalathar) [inline compiler/rustc_mir_transform/src/inline.rs:45]: I would like to suggest another change to this comment, to make it more understandable to other maintainers who don't necessarily know what MC/DC is:

```rust
        // FIXME(#127234): Coverage instrumentation currently doesn't handle inlined
        // MIR correctly when Modified Condition/Decision Coverage is enabled.
```

(Sorry for the inconvenience! I edited my old suggestion shortly after writing it, but it looks like you copied the old version and didn't see the new version.)

Comment (ZhuUx) [reply compiler/rustc_mir_transform/src/inline.rs]: FIXED. Sorry for my carelessness :relieved:

**Review by davidtwco — APPROVED** [review]
This looks good to me. r=me but would like to confirm that @Zalathar is happy with it, knowing coverage more than I do.

**Comment by Zalathar** [comment]
@davidtwco Yes, r+ from me also. 👍

**Comment by ZhuUx** [comment]
No change in the force pull. I just rebase it to the latest master.

**Pull request merged**